spec/UserDbSpec.js

```diff
@@ -13,45 +13,97 @@ describe("User DB", () => {
+        expect(result1)
+            .toBe(true);
+        expect(result2)
+            .toBe(false);
+        expect(result3)
+            .toBe(false);
     });
 
     /* A specification */
-    it("shall be able to add a user with a new username", async function() {
+    it("shall be able to add a user with a new username", async function () {
         let result1 = await UserDb.addUser("zlmonroe", "1234");
         let result1p2 = await UserDb.userLogin("zlmonroe", "1234");
 
-        expect(result1 && result1p2).toBe(true);
+        expect(result1 && result1p2)
+            .toBe(true);
 
         let result3 = await UserDb.addUser("zlmonroe1", "1234");
         let result3p2 = await UserDb.userLogin("zlmonroe1", "1234");
-        expect(result3 && result3p2).toBe(true);
+        expect(result3 && result3p2)
+            .toBe(true);
     });
 
     /* A specification */
-    it("shall fail at adding a user with an existing username", async function() {
+    it("shall fail at adding a user with an existing username", async function () {
         let result1 = await UserDb.addUser("jebush", "1234");
         let result1p2 = await UserDb.userLogin("jebush", "1234");
 
-        expect(result1 && result1p2).toBe(true);
+        expect(result1 && result1p2)
+            .toBe(true);
 
         let result2 = await UserDb.addUser("jebush", "12345");
         let result2p2 = await UserDb.userLogin("jebush", "12345");
-        expect(result2).toBe(false);
-        expect(result2p2).toBe(false);
+        expect(result2)
+            .toBe(false);
+        expect(result2p2)
+            .toBe(false);
 
         let result3 = await UserDb.addUser("jebush1", "1234");
         let result3p2 = await UserDb.userLogin("jebush1", "1234");
-        expect(result3 && result3p2).toBe(true);
+        expect(result3 && result3p2)
+            .toBe(true);
+    });
+
+    /* A specification */
+    it("shall return count of all students in the db", async function () {
+        await UserDb.addUser("test1", "test1");
+        await UserDb.addUser("test2", "test2", "student");
+        await UserDb.addUser("test3", "test3", "teacher");
+
+        const result1 = await UserDb.getUserCount();
+        const result2 = await UserDb.getUserCount("student");
+
+
+        expect(result1)
+            .toBe(result2);
+    });
+
+    /* A specification */
+    it("shall return selected student in the db", async function () {
+        await UserDb.addUser("test1", "test1");
+        await UserDb.addUser("test2", "test2", "student");
+        await UserDb.addUser("test3", "test3", "teacher");
+
+        const result = await UserDb.getUser("test1");
+        expect(result.some(item => item.username === "test1"))
+            .toBe(true);
+    });
+
+    /* A specification */
+    it("shall return all students in the db", async function () {
+        const result = await UserDb.getUsers();
+
+        expect(result.every(item => item.role === "student"))
```

Comment:
This does literally nothing. It iterates through a list of length 0 and checks that all of them have the role student. Not only does checking that all of them have the role student not fulfill this test (i.e. if there were 3 students but only 2 were returned it would pass) currently there are no students for it to even iterate over.

Reply:
Again, I am gonna need some help with these ones because for some reason, the users being returned are different every time because of: 
1. Existing data in the DB before the test starts
2. Some data is removed and added after every test

A potential idea may be to separate the DB for tests, potentially creating a datastore just for the tests. That may be helpful for these ones.

Reply:
I would solve this by doing the following:
1. Create a "unit test db" in the beforeEach setup function in jasmine
2. Clear the unit test db after each function using the cleanup (afterEach type function)

Alternatively, have each set it up for themselves if that works better (just less code efficient if we can do it the other way that is preferred).

Reply:
The second part of my original statement was not resolved:
Not only does checking that all of them have the role student not fulfill this test (i.e. if there were 3 students but only 2 were returned it would pass) currently there are no students for it to even iterate over.

Reply:
You could do this by adding an assert that the length of the students list is 3. Since that would mean that there are 3 users and all are students which matches what you asked.